### PR TITLE
[MRG] Fix net.add_bursty_input() weight updating

### LINKED ISSUE #349: net.add_bursty_drive() uses the same parameter values for all cell types
@cjayb I just found a major issue with `net.add_bursty_drive()`. It seems that only the last value in the weights/delays dictionaries are applied to each cell type. Here's a minimal example to reproduce:
```py 
import os.path as op
import hnn_core
from hnn_core import read_params, default_network, simulate_dipole

hnn_core_root = op.dirname(hnn_core.__file__)
params_fname = op.join(hnn_core_root, 'param', 'default.json')
params = read_params(params_fname)

net = default_network(params)

# Only the last value is stored in the drive conn
weights_ampa = {'L2_basket': 1, 'L2_pyramidal': 3,
                   'L5_basket': 2, 'L5_pyramidal': 4}

syn_delays = {'L2_basket': 0.1, 'L2_pyramidal': 0.1,
                 'L5_basket': 1.0, 'L5_pyramidal': 203423.0}
net.add_bursty_drive(
    'ex_drive', tstart=10, burst_rate=1e-10, burst_std=20., numspikes=2,
    spike_isi=10, repeats=10, location='proximal',
    weights_ampa=weights_ampa, synaptic_delays=syn_delays, seedcore=14)

for type_name, drive in net.external_drives['ex_drive']['conn'].items():
    print(type_name, drive['ampa'])
```
image

This does not impact the evoked or poisson drives. You'll note in `plot_simulate_alpha.py` that the same weights are used for all cell types which is possibly why we didn't notice this.

### PR DESCRIPTION
**Pull request by ntolley**
Closes #349. Turns out that it was a problem with setting `cell_specific=False` in `net._create_drive_conns`, Since the `drive_conn` dictionary is created prior to looping over cell types, all of the cell types share the same dictionary. Just added a `.copy()` to resolve the issue.

### CODE REVIEW

**Comment by ntolley** [comment]
Nevermind, I just realized the fix is more complicated since the shared dictionary is necessary for all cell types to have the same events times when `cell_specific=False`.

I'll come back to this tomorrow, open to any suggestions on how to tackle this. I tried modifying `_instantiate_drives` but no success yet.

**Comment by jasmainak** [comment]
oops ... I think you have the right idea about adding a test first though :) Becoming a proper open source software engineer!

File: hnn_core/network.py [hunk 1]
```diff
@@ -825,7 +825,7 @@ def _create_drive_conns(self, target_populations, weights_by_receptor,
             for cellname in target_populations:
                 drive_conn['target_gids'].extend(self.gid_ranges[cellname])
                 drive_conn['target_type'] = cellname
-                drive_conn_by_cell[cellname] = drive_conn
+                drive_conn_by_cell[cellname] = drive_conn.copy()
```
Comment (cjayb) [inline hnn_core/network.py:828]: This is very insidious, I'm very surprised that the lines below this end up referencing the same object! But I can of course confirm that that's what happens.

File: hnn_core/tests/test_drives.py [hunk 2]
```diff
@@ -117,6 +117,19 @@ def test_add_drives():
     net.add_evoked_drive('early_distal', mu=10, sigma=1, numspikes=1,
                          location='distal')
 
+    # Ensure weights and delays are updated
+    weights_ampa = {'L2_basket': 1.0, 'L2_pyramidal': 3.0,
+                    'L5_basket': 2.0, 'L5_pyramidal': 4.0}
+    syn_delays = {'L2_basket': 1.0, 'L2_pyramidal': 2.0,
+                  'L5_basket': 3.0, 'L5_pyramidal': 4.0}
```
Comment (cjayb) [inline hnn_core/tests/test_drives.py:124]: add these to `test_network` too?

**Comment by cjayb** [comment]
I don't know what to think of the error you're getting because this is a "non-specific" drive. Please keep me posted on what you find: perhaps something can be simplified now that we don't have to slavishly reproduce the GUI anymore?

**Comment by ntolley** [comment]
Making some progress here. I'll need to go through the logic of the fixes a bit more thoroughly, but essentially I added a check to ensure that each target cell type is treated independently.

On the current master branch, specifying a burst input will target ALL cells, even if you exclude them from the `ampa_weights` dict. Here's an example from `test_network.py` to demonstrate:
```py
from copy import deepcopy
import hnn_core
from hnn_core import read_params, default_network
from hnn_core.network_builder import NetworkBuilder
import os.path as op
hnn_core_root = op.dirname(hnn_core.__file__)
params_fname = op.join(hnn_core_root, 'param', 'default.json')
params = read_params(params_fname)
# add rhythmic inputs (i.e., a type of common input)
params.update({'input_dist_A_weight_L2Pyr_ampa': 5.4e-5,
                'input_dist_A_weight_L5Pyr_ampa': 5.4e-5,
                't0_input_dist': 50,
                'input_prox_A_weight_L2Pyr_ampa': 5.4e-5,
                'input_prox_A_weight_L5Pyr_ampa': 5.4e-5,
                't0_input_prox': 50})
net = default_network(deepcopy(params), add_drives_from_params=True)
net.connectivity
```
image


This is meant to instantiate a bursty proximal and distal drive for only the L2 and L5 pyramidal cells. But inspecting `net.connectivity` reveals that connections are made (with identical parameters) for all cells. This means that there are 4 connections made that shouldn't be there (hence the change [here](https://github.com/ntolley/hnn-core/blob/1ac6c1790dd97b0352391bf7ae0879de8c36c2cb/hnn_core/tests/test_network.py#L252)).

**Comment by ntolley** [comment]
Unfortunately these changes are requiring me to update tests which assume a specific structure of bursty inputs. I need to meditate on the logic for this a bit to make sure I am confident in the tests that need changing.

File: hnn_core/network.py [hunk 3]
```diff
@@ -870,17 +870,26 @@ def _instantiate_drives(self, n_trials=1):
...
+                        for drive_cell_gid in this_cell_drive_conn['src_gids']:
+                            event_times.append(_drive_cell_event_times(
+                                drive['type'], this_cell_drive_conn,
+                                drive['dynamics'], trial_idx=trial_idx,
+                                drive_cell_gid=drive_cell_gid,
+                                seedcore=drive['seedcore'])
+                            )
+                else:
+                    for this_cell_drive_conn in drive['conn'].values():
+                        for drive_cell_gid in this_cell_drive_conn['src_gids']:
```
Comment (ntolley) [inline hnn_core/network.py:884]: The inner loop isn't really a for loop because there should only be one `src_gid` when `cell_specific=False`.

File: hnn_core/network.py [hunk 4]
```diff
@@ -870,17 +870,26 @@ def _instantiate_drives(self, n_trials=1):
...
+                            event_ts = _drive_cell_event_times(
+                                drive['type'], this_cell_drive_conn,
+                                drive['dynamics'], trial_idx=trial_idx,
+                                drive_cell_gid=drive_cell_gid,
+                                seedcore=drive['seedcore'])
```
Comment (ntolley) [inline hnn_core/network.py:889]: Since each `this_cell_drive_conn` has the same `seedcore`, they end up getting the same event times. Perhaps there's a cleaner way to only run this once...

File: hnn_core/network.py [hunk 5]
```diff
@@ -870,17 +870,26 @@ def _instantiate_drives(self, n_trials=1):
...
+                            if event_ts:
```
Comment (ntolley) [inline hnn_core/network.py:890]: Necessary when some cells have empty weights. Otherwise the shape of the list won't match the number of cells being targeted.

**Comment by jasmainak** [comment]
@ntolley you might be interested in this commit:

https://github.com/jonescompneurolab/hnn-core/pull/221/commits/5809750020bd14eb0dbb2aabcc8e1548f9843cdb

I thought I was making logically equivalent code but maybe I introduced a bug while doing that

**Comment by ntolley** [comment]
They weren't kidding about the dragons 🐲!  As far as I can tell, that change didn't introduce any problems. 

I think I'm getting close to resolving this issue, but the alpha example may need some tweaking since the drives are now different.

File: hnn_core/network.py [hunk 6]
```diff
@@ -870,17 +870,37 @@ def _instantiate_drives(self, n_trials=1):
...
+                                drive['type'], this_cell_drive_conn,
+                                drive['dynamics'], trial_idx=trial_idx,
+                                drive_cell_gid=drive_cell_gid,
+                                seedcore=drive['seedcore'])
+                            )
+                else:
+                    for this_cell_drive_conn in drive['conn'].values():
+                        has_weight = np.logical_or(
+                            this_cell_drive_conn['ampa'],
+                            this_cell_drive_conn['nmda'])
```
Comment (jasmainak) [inline hnn_core/network.py:886]: isn't this logic already in `_drive_cell_event_times` ?

https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/drives.py#L237

Comment (ntolley) [reply hnn_core/network.py:886]: Ah good point. I think both are necessary. For `cell_specific=True`, this returns an empty list for cells with weight=0.

However, for `cell_specific=False`, you need to be careful because all cells are sharing the same event times. Sometimes the parameters of the drives lead to an empty list, despite the cell weight being non-zero. This ensures that if an empty list is returned, it is due to the simulation parameters and not the weights.

Comment (ntolley) [reply hnn_core/network.py:886]: Said another way, the loop exits the first time ` _drive_cell_event_times` is called so that there is only one set of event_times. You need to be sure the function receives a drive_conn with positive weights.

**Comment by ntolley** [comment]
I think the problem has been fixed @cjayb @jasmainak! 

Once the circleci artifact is generated I'll push some doc updates. Hopefully the alpha/beta example is roughly similar.

**Comment by ntolley** [comment]
This stumped me for a moment, there is virtually no difference between the new and the old alpha beta example. But inspecting `net.connectivity` easily confirms that the current master branch produces more connections than there should be (and with the wrong weights/delays). 

Since this is a sub-threshold  phenomenon, it actually makes sense that spurious basket cell connections don't impact the current dipole. However, I am surprised that have a different delay between the pyramidal cells (0.1 vs. 1.0) does not impact the alpha rhythm (compared to master branch where both delays are set to 1.0). Perhaps the delay is negligible with the ongoing issues with cell spacing.

Nevertheless, #348 should take full advantage of this PR since it completely relies on non-uniform drive weights. It is actually how I stumbled on this bug. I realized no matter how high I set the ampa weight on the basket cells, they simply would not fire.

**Review by jasmainak — APPROVED** [review]

**Comment by jasmainak** [comment]
@ntolley see the commits I pushed and if you agree with them

I'd take another set of reviews from @cjayb since he wrote the original code

**Comment by jasmainak** [comment]
@cjayb merge button is yours once you're happy

File: hnn_core/tests/test_network.py [hunk 7]
```diff
@@ -17,11 +17,11 @@ def test_network():
...
     # add rhythmic inputs (i.e., a type of common input)
-    params.update({'input_dist_A_weight_L2Pyr_ampa': 5.4e-5,
-                   'input_dist_A_weight_L5Pyr_ampa': 5.4e-5,
+    params.update({'input_dist_A_weight_L2Pyr_ampa': 1.4e-5,
+                   'input_dist_A_weight_L5Pyr_ampa': 2.4e-5,
                    't0_input_dist': 50,
-                   'input_prox_A_weight_L2Pyr_ampa': 5.4e-5,
-                   'input_prox_A_weight_L5Pyr_ampa': 5.4e-5,
+                   'input_prox_A_weight_L2Pyr_ampa': 3.4e-5,
+                   'input_prox_A_weight_L5Pyr_ampa': 4.4e-5,
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:24]: why were these changed @ntolley ? or did I mess up somewhere with the rebase?

Comment (ntolley) [reply hnn_core/tests/test_network.py:24]: This just reflects that the bug can result in all cells taking the same weights when they share a dictionary.

I think the reason we didn't notice this bug in the first place is because we often used the same weights for all cells when adding a bursty input.

File: hnn_core/network.py [hunk 8]
```diff
@@ -869,20 +869,33 @@ def _instantiate_drives(self, n_trials=1):
...
+                    if drive['cell_specific']:
+                        for drive_cell_gid in this_cell_drive_conn['src_gids']:
+                            event_times.append(_drive_cell_event_times(
+                                drive['type'], this_cell_drive_conn,
+                                drive['dynamics'], trial_idx=trial_idx,
+                                drive_cell_gid=drive_cell_gid,
+                                seedcore=drive['seedcore'])
+                            )
+                    else:
+                        # cell_specific=False should only have one src_gid
```
Comment (ntolley) [inline hnn_core/network.py:884]: I'm not sure these changes make sense, if `cell_specific=False` the events are never appended to `event_list`

File: hnn_core/network.py [hunk 9]
```diff
@@ -869,20 +869,33 @@ def _instantiate_drives(self, n_trials=1):
...
+                        assert len(this_cell_drive_conn['src_gids']) == 1
+
```
Comment (ntolley) [inline hnn_core/network.py:886]:
```suggestion
# Only return empty event times if all cells have no events
```

Comment (ntolley) [reply hnn_core/network.py:886]: I get the feeling we may need a "here there be dragons" message of our own as this code is dealing with edge cases that are not immediately obvious (empty events due to 0 weight, or the parameters of the drive)

Comment (jasmainak) [reply hnn_core/network.py:886]: done, added this comment. By the way, I was surprised that this part of the code was not tested directly. Moving forward, we should add tests that check these corner cases

**Review by cjayb — APPROVED** [review]
Excellent, thanks for catching that @ntolley
